compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -6449,8 +6449,19 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
   }
 
   /** Test whether `sym` is the symbol of a JS function definition */
-  private def isJSFunctionDef(sym: Symbol): Boolean =
-    sym.isAnonymousClass && AllJSFunctionClasses.exists(sym isSubClass _)
+  private def isJSFunctionDef(sym: Symbol): Boolean = {
+    /* A JS function may only reach us in the backend, if it originally was a
+     * lambda. This is difficult to check in the backend, so we use the fact
+     * that a non-lambda, concrete, non-native JS type, cannot implement a
+     * method named `apply`.
+     *
+     * All JS function classes have an abstract member named `apply`. Therefore,
+     * a class is a JS lambda iff it is concrete, a non-native JS type and
+     * inherits from a JS function class.
+     */
+    !isJSNativeClass(sym) && !sym.isAbstract &&
+    AllJSFunctionClasses.exists(sym.isSubClass(_))
```

Comment:
The new test means that we'll have to go through all the `AllJSFunctionClasses` to test `isSubClass(_)` for most input `sym`s. This may have a non trivial performance impact.

I think we could keep the check `sym.isAnonymousClass &&` as a first fast-path test. That would avoid the expensive tests for most classes, I believe.